src/manifold/src/smoothing.cpp

```diff
@@ -501,6 +501,269 @@ class Partition {
 
 namespace manifold {
 
+// sharpenedEdges are referenced to the input Mesh, but the triangles have
+// been sorted in creating the Manifold, so the indices are converted using
+// meshRelation_.
+std::vector<Smoothness> Manifold::Impl::UpdateSharpenedEdges(
+    const std::vector<Smoothness>& sharpenedEdges) const {
+  std::unordered_map<int, int> oldHalfedge2New;
+  for (int tri = 0; tri < NumTri(); ++tri) {
+    int oldTri = meshRelation_.triRef[tri].tri;
+    for (int i : {0, 1, 2}) oldHalfedge2New[3 * oldTri + i] = 3 * tri + i;
+  }
+  std::vector<Smoothness> newSharp = sharpenedEdges;
+  for (Smoothness& edge : newSharp) {
+    edge.halfedge = oldHalfedge2New[edge.halfedge];
+  }
+  return newSharp;
+}
+
+// Find faces containing at least 3 triangles - these will not have
+// interpolated normals - all their vert normals must match their face normal.
+Vec<bool> Manifold::Impl::FlatFaces() const {
+  const int numTri = NumTri();
+  Vec<bool> triIsFlatFace(numTri, false);
+  for (int tri = 0; tri < numTri; ++tri) {
+    const TriRef& ref = meshRelation_.triRef[tri];
+    int faceNeighbors = 0;
+    glm::ivec3 faceTris = {-1, -1, -1};
+    for (const int j : {0, 1, 2}) {
+      const int neighborTri =
+          halfedge_[halfedge_[3 * tri + j].pairedHalfedge].face;
+      const TriRef& jRef = meshRelation_.triRef[neighborTri];
+      if (jRef.SameFace(ref)) {
+        ++faceNeighbors;
+        faceTris[j] = neighborTri;
+      }
+    }
+    if (faceNeighbors > 1) {
+      triIsFlatFace[tri] = true;
+      for (const int j : {0, 1, 2}) {
+        if (faceTris[j] >= 0) {
+          triIsFlatFace[faceTris[j]] = true;
+        }
+      }
+    }
+  }
+  return triIsFlatFace;
+}
+
+// Returns a vector of length numVert that has a tri that is part of a
+// neighboring flat face if there is only one flat face. If there are none it
+// gets -1, and if there are more than one it gets -2.
+Vec<int> Manifold::Impl::VertFlatFace(const Vec<bool>& flatFaces) const {
+  Vec<int> vertFlatFace(NumVert(), -1);
+  Vec<TriRef> vertRef(NumVert(), {-1, -1, -1});
+  for (int tri = 0; tri < NumTri(); ++tri) {
+    if (flatFaces[tri]) {
+      for (const int j : {0, 1, 2}) {
+        const int vert = halfedge_[3 * tri + j].startVert;
+        if (vertRef[vert].SameFace(meshRelation_.triRef[tri])) continue;
+        vertRef[vert] = meshRelation_.triRef[tri];
+        vertFlatFace[vert] = vertFlatFace[vert] == -1 ? tri : -2;
+      }
+    }
+  }
+  return vertFlatFace;
+}
+
+std::vector<Smoothness> Manifold::Impl::SharpenEdges(
+    float minSharpAngle, float minSmoothness) const {
+  std::vector<Smoothness> sharpenedEdges;
+  for (int e = 0; e < halfedge_.size(); ++e) {
+    if (!halfedge_[e].IsForward()) continue;
+    const int pair = halfedge_[e].pairedHalfedge;
+    const float dihedral = glm::degrees(
```

Comment:
maybe it will be slightly faster if we convert the angle into radian and only compare using radian, instead of doing conversion of degree to radian in a loop. same for the dihedral calculation below as well.